File: problems/46. Permutations/main.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        path = []
        used = [False] * len(nums)

        def helper():
            if len(path) == len(nums):
                result.append(list(path))
                return
            for i in range(len(nums)):
                if not used[i]:
                    used[i] = True
                    path.append(nums[i])
                    helper()
                    path.pop()
                    used[i] = False

        helper()
        return result

File: problems/46. Permutations/main.go
func permute(nums []int) [][]int {
	var result [][]int
	var path []int
	used := make([]bool, len(nums))

	var helper func()
	helper = func() {
		if len(path) == len(nums) {
			permutation := append([]int{}, path...)
			result = append(result, permutation)
			return
		}
		for i := 0; i < len(nums); i++ {
			if !used[i] {
				used[i] = true
				path = append(path, nums[i])
				helper()
				path = path[:len(path)-1]
				used[i] = false
			}
		}
	}

	helper()
	return result
}
